packages/core/admin/server/src/services/homepage.ts

```diff
@@ -0,0 +1,131 @@
+        where: {
+          role: { users: { id: strapi.requestContext.get()?.state?.user.id } },
+          action: 'plugin::content-manager.explorer.read',
+        },
+      });
+      const allowedContentTypeNames = readPermissions
+        .map((permission) => permission.subject)
+        .filter(Boolean) as UID.ContentType[];
+
+      // Fetch the configuration for each content type in a single query
+      const rawConfigurations = await coreStore.findMany({
+        where: {
+          key: {
+            $in: allowedContentTypeNames.map(
+              (contentType) => `plugin_content_manager_configuration_content_types::${contentType}`
+            ),
+          },
+        },
+      });
+      const configurations = rawConfigurations.map((rawConfiguration) => {
+        return JSON.parse(rawConfiguration.value);
+      });
+
+      const recentDocuments = await Promise.all(
+        allowedContentTypeNames.map(async (contentTypeName) => {
+          const configuration = configurations.find((config) => config.uid === contentTypeName);
+          const contentType = strapi.contentType(contentTypeName);
+          const fields = ['documentId', 'updatedAt'];
+
+          // Add fields required to get the status if D&P is enabled
+          const hasDraftAndPublish = contentTypes.hasDraftAndPublish(contentType);
+          if (hasDraftAndPublish) {
+            fields.push('publishedAt');
+          }
+
+          // Only add the main field if it's defined
+          const { mainField } = configuration.settings;
+          if (mainField) {
+            fields.push(mainField);
+          }
+
+          // Only add locale if it's localized
+          const isLocalized = (contentType.pluginOptions?.i18n as any)?.localized;
+          if (isLocalized) {
+            fields.push('locale');
+          }
+
+          const documents = await strapi.documents(contentTypeName).findMany({
+            limit: MAX_DOCUMENTS,
+            sort: 'updatedAt:desc',
+            fields,
+          });
+
+          return documents.map((document) => {
+            /**
+             * Save the main field value before deleting it so we can use the common
+             * title key instead across all content types. Use the delete operator instead of
+             * destructuring or lodash omit for better type inference.
+             */
+            const mainFieldValue = document[mainField ?? 'documentId'];
+            delete document[mainField];
+
+            return {
+              data: {
+                ...document,
+                updatedAt: new Date(document.updatedAt),
+                title: mainFieldValue,
+              },
+              meta: {
+                model: contentTypeName,
+                kind: contentType.kind,
+                hasDraftAndPublish,
+              },
+            };
+          });
+        })
+      );
+
+      const overallRecentDocuments = recentDocuments
+        .flat()
```

Comment:
I think the nested loop and flattening could be avoided by splitting into two arrays. One to compute the info you need with fields. A second one that maps the first one to do the actual async stuff.

Reply:
Not sure why Github didn't include my entire selection but I'm talking about everything inside the Promise.all above